File: impl/go/utilities.go
package twill

import "strings"

// CompileStatus is the outcome of a utility definition (SPEC §4.1.7).
type CompileStatus int

const (
	// NotHandled means this definition does not handle the candidate.
	NotHandled CompileStatus = iota
	// Handled means success.
	Handled
	// Invalid means the candidate is invalid for this definition.
	Invalid
)

// CompileFn compiles a candidate.
type CompileFn func(c *Candidate) ([]Node, CompileStatus)

// UtilityDefinition is a registered utility.
type UtilityDefinition struct {
	Kind    UtilityKind
	Compile CompileFn
	// Types marks a fallback definition when it has more than one entry and
	// includes "any".
	Types []string
}

// IsFallback reports whether the definition is a fallback definition.
func (d *UtilityDefinition) IsFallback() bool {
	if len(d.Types) <= 1 {
		return false
	}
	for _, t := range d.Types {
		if t == "any" {
			return true
		}
	}
	return false
}

// Utilities is the utility registry (SPEC §10.1).
type Utilities struct {
	defs  map[string][]*UtilityDefinition
	names []string
}

// NewUtilities creates an empty registry.
func NewUtilities() *Utilities { return &Utilities{defs: map[string][]*UtilityDefinition{}} }

func (u *Utilities) add(name string, d *UtilityDefinition) {
	if _, ok := u.defs[name]; !ok {
		u.names = append(u.names, name)
	}
	u.defs[name] = append(u.defs[name], d)
}

// Static registers a static definition.
func (u *Utilities) Static(name string, compile CompileFn) {
	u.add(name, &UtilityDefinition{Kind: UtilityStatic, Compile: compile})
}

// Functional registers a functional definition.
func (u *Utilities) Functional(name string, compile CompileFn, types []string) {
	u.add(name, &UtilityDefinition{Kind: UtilityFunctional, Compile: compile, Types: types})
}

// Has reports whether a definition of the kind exists for name.
func (u *Utilities) Has(name string, kind UtilityKind) bool {
	for _, d := range u.defs[name] {
		if d.Kind == kind {
			return true
		}
	}
	return false
}

// Get returns the definitions for name.
func (u *Utilities) Get(name string) []*UtilityDefinition { return u.defs[name] }

// Names returns the registered names.
func (u *Utilities) Names() []string { return append([]string{}, u.names...) }

// AsColor applies a modifier to a color value (SPEC §10.3).
func AsColor(value string, modifier *Modifier, theme *Theme) (string, bool) {
	if modifier == nil {
		return value, true
	}
	if modifier.Kind == ValueArbitrary {
		return WithAlpha(value, modifier.Value), true
	}
	if opacity, ok := theme.Resolve(strPtr(modifier.Value), []string{"--opacity"}, 0); ok {
		return WithAlpha(value, opacity), true
	}
	if !IsMultipleOfQuarter(modifier.Value) {
		return "", false
	}
	return WithAlpha(value, modifier.Value+"%"), true
}

// ResolveThemeColor resolves a named color candidate value through the theme.
func ResolveThemeColor(c *Candidate, themeKeys []string, theme *Theme) (string, bool) {
	if c.Value == nil || c.Value.Kind != ValueNamed {
		return "", false
	}
	var value string
	switch c.Value.Value {
	case "inherit":
		value = "inherit"
	case "transparent":
		value = "transparent"
	case "current":
		value = "currentcolor"
	default:
		v, ok := theme.Resolve(strPtr(c.Value.Value), themeKeys, 0)
		if !ok {
			return "", false
		}
		value = v
	}
	return AsColor(value, c.Modifier, theme)
}

// Declarations is a list of (property, value) pairs.
type Declarations [][2]string

// StaticUtility registers a static definition returning the declarations.
func StaticUtility(u *Utilities, name string, declarations Declarations) {
	u.Static(name, func(*Candidate) ([]Node, CompileStatus) {
		nodes := make([]Node, 0, len(declarations))
		for _, d := range declarations {
			nodes = append(nodes, Decl(d[0], d[1]))
		}
		return nodes, Handled
	})
}

// StaticUtilityFn registers a static definition backed by a function.
func StaticUtilityFn(u *Utilities, name string, fn func() []Node) {
	u.Static(name, func(*Candidate) ([]Node, CompileStatus) { return fn(), Handled })
}

// FunctionalUtilityDescription describes a functional utility (SPEC §10.2).
type FunctionalUtilityDescription struct {
	SupportsNegative  bool
	SupportsFractions bool
	ThemeKeys         []string
	// DefaultValue is the value for a candidate without a value segment.
	// NoDefault marks the null case (no output); when both are unset the
	// first namespace itself is resolved.
	DefaultValue            *string
	NoDefault               bool
	StaticValues            map[string][]Node
	HandleBareValue         func(v *CandidateValue) (string, bool)
	HandleNegativeBareValue func(v *CandidateValue) (string, bool)
	Handle                  func(value string, dataType *string) []Node
}

// FunctionalUtility registers a functional definition (SPEC §10.2).
func FunctionalUtility(u *Utilities, theme *Theme, root string, desc FunctionalUtilityDescription) {
	compile := func(c *Candidate, negative bool) ([]Node, CompileStatus) {
		var value string
		resolved := false
		var dataType *string

		switch {
		case c.Value == nil:
			if c.Modifier != nil {
				return nil, NotHandled
			}
			switch {
			case desc.NoDefault:
			case desc.DefaultValue != nil:
				value, resolved = *desc.DefaultValue, true
			default:
				value, resolved = theme.Resolve(nil, desc.ThemeKeys, 0)
			}
		case c.Value.Kind == ValueArbitrary:
			if c.Modifier != nil {
				return nil, NotHandled
			}
			value, resolved = c.Value.Value, true
			dataType = c.Value.DataType
		default:
			named := c.Value
			fractionConsumed := false
			if named.Fraction != nil {
				if v, ok := theme.Resolve(named.Fraction, desc.ThemeKeys, 0); ok {
					value, resolved, fractionConsumed = v, true, true
				}
			}
			if !resolved {
				if v, ok := theme.Resolve(strPtr(named.Value), desc.ThemeKeys, 0); ok {
					if c.Modifier != nil && !fractionConsumed {
						return nil, NotHandled
					}
					value, resolved = v, true
				}
			}
			if !resolved && desc.SupportsFractions && named.Fraction != nil {
				parts := strings.SplitN(*named.Fraction, "/", 2)
				if !IsPositiveInteger(parts[0]) || !IsPositiveInteger(parts[1]) {
					return nil, NotHandled
				}
				value, resolved = "calc("+parts[0]+" / "+parts[1]+" * 100%)", true
			}
			if !resolved && negative && desc.HandleNegativeBareValue != nil {
				if bare, ok := desc.HandleNegativeBareValue(named); ok {
					if !strings.Contains(bare, "/") && c.Modifier != nil {
						return nil, NotHandled
					}
					return desc.Handle(bare, nil), Handled
				}
			}
			if !resolved && desc.HandleBareValue != nil {
				if bare, ok := desc.HandleBareValue(named); ok {
					if !strings.Contains(bare, "/") && c.Modifier != nil {
						return nil, NotHandled
					}
					value, resolved = bare, true
				}
			}
			if !resolved && !negative && c.Modifier == nil && desc.StaticValues != nil {
				if nodes, ok := desc.StaticValues[named.Value]; ok {
					return CloneNodes(nodes), Handled
				}
			}
		}

		if !resolved {
			return nil, NotHandled
		}
		if negative {
			value = "calc(" + value + " * -1)"
		}
		nodes := desc.Handle(value, dataType)
		if nodes == nil {
			return nil, NotHandled
		}
		return nodes, Handled
	}

	u.Functional(root, func(c *Candidate) ([]Node, CompileStatus) { return compile(c, false) }, nil)
	if desc.SupportsNegative {
		u.Functional("-"+root, func(c *Candidate) ([]Node, CompileStatus) { return compile(c, true) }, nil)
	}
}

// ColorUtility registers a functional definition that resolves its value as a color.
func ColorUtility(u *Utilities, theme *Theme, root string, themeKeys []string, handle func(value string) []Node) {
	u.Functional(root, func(c *Candidate) ([]Node, CompileStatus) {
		if c.Value == nil {
			return nil, NotHandled
		}
		var value string
		var ok bool
		if c.Value.Kind == ValueArbitrary {
			value, ok = AsColor(c.Value.Value, c.Modifier, theme)
		} else {
			value, ok = ResolveThemeColor(c, themeKeys, theme)
		}
		if !ok {
			return nil, NotHandled
		}
		return handle(value), Handled
	}, nil)
}

// SpacingUtility registers `<name>-px`, optionally `-<name>-px`, and a
// functional utility whose bare values are spacing multipliers.
func SpacingUtility(u *Utilities, theme *Theme, name string, themeKeys []string, handle func(value string) []Node, supportsNegative, supportsFractions bool) {
	StaticUtilityFn(u, name+"-px", func() []Node { return handle("1px") })
	if supportsNegative {
		StaticUtilityFn(u, "-"+name+"-px", func() []Node { return handle("-1px") })
	}
	spacing := func(prefix string) func(v *CandidateValue) (string, bool) {
		return func(v *CandidateValue) (string, bool) {
			if !IsMultipleOfQuarter(v.Value) {
				return "", false
			}
			if _, ok := theme.Resolve(nil, []string{"--spacing"}, 0); !ok {
				return "", false
			}
			return "--spacing(" + prefix + v.Value + ")", true
		}
	}
	FunctionalUtility(u, theme, name, FunctionalUtilityDescription{
		ThemeKeys:               themeKeys,
		NoDefault:               true,
		SupportsNegative:        supportsNegative,
		SupportsFractions:       supportsFractions,
		HandleBareValue:         spacing(""),
		HandleNegativeBareValue: spacing("-"),
		Handle:                  func(value string, _ *string) []Node { return handle(value) },
	})
}

// BareInteger resolves a bare value when it is a positive integer.
func BareInteger(v *CandidateValue) (string, bool) {
	return v.Value, IsPositiveInteger(v.Value)
}

File: impl/ocaml/src/utilities.ml
(* The utility registry and the utility-definition helpers (SPEC §4.1.7, §10.1 through §10.3). *)

open Ast
open Candidate

type compile_status = Not_handled | Handled | Invalid
type compile_fn = candidate -> node list * compile_status

type definition = {
  ukind : utility_kind;
  compile : compile_fn;
  (* Marks a fallback definition when it has more than one entry including "any". *)
  types : string list option;
}

let is_fallback d = match d.types with Some ts when List.length ts > 1 -> List.mem "any" ts | _ -> false

type t = { defs : (string, definition list) Hashtbl.t; mutable names_rev : string list }

let create () = { defs = Hashtbl.create 1024; names_rev = [] }

let add u name d =
  match Hashtbl.find_opt u.defs name with
  | Some existing -> Hashtbl.replace u.defs name (existing @ [ d ])
  | None ->
      u.names_rev <- name :: u.names_rev;
      Hashtbl.replace u.defs name [ d ]

let static u name compile = add u name { ukind = Static; compile; types = None }
let functional u ?types name compile = add u name { ukind = Functional; compile; types }
let has u name kind = match Hashtbl.find_opt u.defs name with Some ds -> List.exists (fun d -> d.ukind = kind) ds | None -> false
let get u name = match Hashtbl.find_opt u.defs name with Some ds -> ds | None -> []
let names u = List.rev u.names_rev

(* Applies a modifier to a color value (SPEC §10.3). *)
let as_color value (modifier : modifier option) theme =
  match modifier with
  | None -> Some value
  | Some { mkind = Arbitrary; mvalue } -> Some (Theme.with_alpha value mvalue)
  | Some { mvalue; _ } -> (
      match Theme.resolve theme (Some mvalue) [ "--opacity" ] 0 with
      | Some opacity -> Some (Theme.with_alpha value opacity)
      | None -> if Utils.is_multiple_of_quarter mvalue then Some (Theme.with_alpha value (mvalue ^ "%")) else None)

(* Resolves a named color candidate value through the theme. *)
let resolve_theme_color (c : candidate) theme_keys theme =
  match c.cvalue with
  | Some { vkind = Named; value; _ } -> (
      let resolved =
        match value with
        | "inherit" -> Some "inherit"
        | "transparent" -> Some "transparent"
        | "current" -> Some "currentcolor"
        | _ -> Theme.resolve theme (Some value) theme_keys 0
      in
      match resolved with Some v -> as_color v c.cmodifier theme | None -> None)
  | _ -> None

let not_handled = ([], Not_handled)
let handled nodes = (nodes, Handled)

(* Registers a static definition returning the declarations. *)
let static_utility u name declarations =
  static u name (fun _ -> handled (List.map (fun (p, v) -> decl p v) declarations))

(* Registers a static definition backed by a function. *)
let static_utility_fn u name f = static u name (fun _ -> handled (f ()))

(* Describes a functional utility (SPEC §10.2). *)
type description = {
  supports_negative : bool;
  supports_fractions : bool;
  theme_keys : string list;
  (* The value for a candidate without a value segment; [no_default] marks the
     null case and when both are unset the first namespace itself resolves. *)
  default_value : string option;
  no_default : bool;
  static_values : (string * node list) list;
  handle_bare_value : (candidate_value -> string option) option;
  handle_negative_bare_value : (candidate_value -> string option) option;
  handle : string -> string option -> node list;
}

let describe ?(supports_negative = false) ?(supports_fractions = false) ?(theme_keys = []) ?default_value
    ?(no_default = false) ?(static_values = []) ?handle_bare_value ?handle_negative_bare_value handle =
  { supports_negative; supports_fractions; theme_keys; default_value; no_default; static_values; handle_bare_value;
    handle_negative_bare_value; handle }

exception Result of (node list * compile_status)

(* Registers a functional definition (SPEC §10.2). *)
let functional_utility u theme root desc =
  let compile (c : candidate) negative =
    try
      let resolved = ref None and data_type = ref None in
      (match c.cvalue with
      | None ->
          if c.cmodifier <> None then raise (Result not_handled);
          if desc.no_default then ()
          else if desc.default_value <> None then resolved := desc.default_value
          else resolved := Theme.resolve theme None desc.theme_keys 0
      | Some { vkind = Arbitrary; value; data_type = dt; _ } ->
          if c.cmodifier <> None then raise (Result not_handled);
          resolved := Some value;
          data_type := dt
      | Some named ->
          let fraction_consumed = ref false in
          (match named.fraction with
          | Some f -> (
              match Theme.resolve theme (Some f) desc.theme_keys 0 with
              | Some v ->
                  resolved := Some v;
                  fraction_consumed := true
              | None -> ())
          | None -> ());
          if !resolved = None then begin
            match Theme.resolve theme (Some named.value) desc.theme_keys 0 with
            | Some v ->
                if c.cmodifier <> None && not !fraction_consumed then raise (Result not_handled);
                resolved := Some v
            | None -> ()
          end;
          if !resolved = None && desc.supports_fractions then begin
            match named.fraction with
            | Some f ->
                let a, b = match String.index_opt f '/' with Some i -> (String.sub f 0 i, Utils.after f (i + 1)) | None -> (f, "") in
                if not (Utils.is_positive_integer a && Utils.is_positive_integer b) then raise (Result not_handled);
                resolved := Some ("calc(" ^ a ^ " / " ^ b ^ " * 100%)")
            | None -> ()
          end;
          if !resolved = None && negative then begin
            match desc.handle_negative_bare_value with
            | Some h -> (
                match h named with
                | Some bare ->
                    if (not (String.contains bare '/')) && c.cmodifier <> None then raise (Result not_handled);
                    raise (Result (handled (desc.handle bare None)))
                | None -> ())
            | None -> ()
          end;
          if !resolved = None then begin
            match desc.handle_bare_value with
            | Some h -> (
                match h named with
                | Some bare ->
                    if (not (String.contains bare '/')) && c.cmodifier <> None then raise (Result not_handled);
                    resolved := Some bare
                | None -> ())
            | None -> ()
          end;
          if !resolved = None && (not negative) && c.cmodifier = None then begin
            match List.assoc_opt named.value desc.static_values with
            | Some nodes -> raise (Result (handled (clone_nodes nodes)))
            | None -> ()
          end);
      match !resolved with
      | None -> not_handled
      | Some value ->
          let value = if negative then "calc(" ^ value ^ " * -1)" else value in
          let nodes = desc.handle value !data_type in
          if nodes = [] then not_handled else handled nodes
    with Result r -> r
  in
  functional u root (fun c -> compile c false);
  if desc.supports_negative then functional u ("-" ^ root) (fun c -> compile c true)

(* Registers a functional definition that resolves its value as a color. *)
let color_utility u theme root theme_keys handle =
  functional u root (fun (c : candidate) ->
      match c.cvalue with
      | None -> not_handled
      | Some { vkind = Arbitrary; value; _ } -> (
          match as_color value c.cmodifier theme with Some v -> handled (handle v) | None -> not_handled)
      | Some _ -> ( match resolve_theme_color c theme_keys theme with Some v -> handled (handle v) | None -> not_handled))

(* Registers `<name>-px`, optionally `-<name>-px`, and a functional utility
   whose bare values are spacing multipliers. *)
let spacing_utility u theme name theme_keys handle supports_negative supports_fractions =
  static_utility_fn u (name ^ "-px") (fun () -> handle "1px");
  if supports_negative then static_utility_fn u ("-" ^ name ^ "-px") (fun () -> handle "-1px");
  let spacing prefix (v : candidate_value) =
    if not (Utils.is_multiple_of_quarter v.value) then None
    else if Theme.resolve theme None [ "--spacing" ] 0 = None then None
    else Some ("--spacing(" ^ prefix ^ v.value ^ ")")
  in
  functional_utility u theme name
    (describe ~theme_keys ~no_default:true ~supports_negative ~supports_fractions ~handle_bare_value:(spacing "")
       ~handle_negative_bare_value:(spacing "-") (fun value _ -> handle value))

(* Resolves a bare value when it is a positive integer. *)
let bare_integer (v : candidate_value) = if Utils.is_positive_integer v.value then Some v.value else None

(* Resolves a bare positive integer with a unit suffix. *)
let bare_suffix suffix (v : candidate_value) = if Utils.is_positive_integer v.value then Some (v.value ^ suffix) else None

(* The explicit or inferred data type of an arbitrary candidate value. *)
let arbitrary_type (c : candidate) types =
  match c.cvalue with
  | Some { data_type = Some t; _ } -> t
  | Some { value; _ } -> Data_types.infer_data_type value types
  | None -> ""
